File: Tests/CaptureTests.cs
using System;
using System.Text.Patterns;
using Microsoft.VisualStudio.TestTools.UnitTesting;

namespace Tests {
	[TestClass]
	public class CaptureTests {
		[TestMethod]
		public void Constructor() {
			Pattern _ = (+Pattern.Letter).Capture(out Capture _);
		}

		[TestMethod]
		public void Consume() {
			Capture Capture;
			Pattern Capturer = (+Pattern.Letter).Capture(out Capture);
			Result Result;

			Result = Capturer.Consume("Hello");
			ResultAssert.Captures("Hello", Result);
			CaptureAssert.Captures("Hello", Capture);

			Pattern Start = "Hello " & (+Pattern.Letter).Capture(out Capture) & "!";

			Result = Start.Consume("Hello World!");
			ResultAssert.Captures("Hello World!", Result);
			CaptureAssert.Captures("World", Capture);

			Pattern Stop = (Pattern)"Goodbye " & Capture & ".";

			Result = Stop.Consume("Goodbye World.");
			ResultAssert.Captures("Goodbye World.", Result);

			Result = Start.Consume("Hello Range!");
			ResultAssert.Captures("Hello Range!", Result);
			CaptureAssert.Captures("Range", Capture);

			Pattern Range = (Start, Stop);

			Result = Range.Consume("Hello Range! How are you today? Goodbye Range. Have a good day.");
			ResultAssert.Captures("Hello Range! How are you today? Goodbye Range.", Result);
			CaptureAssert.Captures("Range", Capture);
		}
	}
}

File: Tests/Patterns/CaptureTests.fs
namespace Tests.Patterns

open System
open Stringier.Patterns
open Microsoft.VisualStudio.TestTools.UnitTesting

[<TestClass>]
type CaptureTests() =

    [<TestMethod>]
    member _.``constructor`` () =
        let mutable capture = null
        let _pattern = ('a' || 'b' || 'c') => &capture
        ()

    [<TestMethod>]
        member _.``consume`` () =
            let mutable capture = Capture()
            let mutable pattern = ('a' || 'b' || 'c') => &capture
            let mutable result = Result()
           
            result <- pattern.Consume("a")
            ResultAssert.Captures("a", result)
            CaptureAssert.Captures("a", capture)

            pattern <- ('a' || 'b' || 'c') => &capture >> '!'

            result <- pattern.Consume("a")
            ResultAssert.Fails(result)

            result <- pattern.Consume("a!")
            ResultAssert.Captures("a!", result)
            CaptureAssert.Captures("a", capture)

            pattern <- ('a' || 'b' || 'c') => &capture >> ',' >> capture